### PR TITLE
[FAB-18509] Stop panic if collection index path is wrong

### PR DESCRIPTION
**Pull request by mbwhite**
Signed-off-by: Matthew B White <whitemat@uk.ibm.com>

#### Type of change
- Bug fix

### CODE REVIEW

File: core/ledger/kvledger/txmgmt/privacyenabledstate/db.go [hunk 1]
```diff
@@ -420,21 +420,26 @@ const (
 	// Example for chaincode indexes:
 	// "META-INF/statedb/couchdb/indexes/indexColorSortName.json"
 	chaincodeIndexDirDepth = 3
+	pathLengthForCCindexes = 5
```
Comment (manish-sethi) [inline core/ledger/kvledger/txmgmt/privacyenabledstate/db.go]:
```suggestion
	pathLengthForCCindexes = 4
```
Path that comes here is of [directory only](https://github.com/hyperledger/fabric/blob/main/core/ledger/kvledger/txmgmt/privacyenabledstate/db.go#L323). The file name is not part of this.

File: core/ledger/kvledger/txmgmt/privacyenabledstate/db.go [hunk 2]
```diff
@@ -420,21 +420,26 @@ const (
...
+
 	// Example for collection scoped indexes:
 	// "META-INF/statedb/couchdb/collections/collectionMarbles/indexes/indexCollMarbles.json"
-	collectionDirDepth      = 3
-	collectionNameDepth     = 4
-	collectionIndexDirDepth = 5
+	collectionDirDepth             = 3
+	collectionNameDepth            = 4
+	collectionIndexDirDepth        = 5
+	pathLengthForCollectionindexes = 7
```
Comment (manish-sethi) [inline core/ledger/kvledger/txmgmt/privacyenabledstate/db.go]:
```suggestion
	pathLengthForCollectionindexes = 6
```
same as above

**Review by manish-sethi — CHANGES_REQUESTED** [review]

File: core/ledger/kvledger/txmgmt/privacyenabledstate/db.go (hunk 1 above)
Comment (mbwhite) [reply core/ledger/kvledger/txmgmt/privacyenabledstate/db.go]: thanks for the review Manish - this is the output I got from the unit test - with printing the args to the `getIndexInfo` function. Includes the index file name
```
=== RUN   TestGetIndexInfo
[getIndexInfo] indexPath=META-INF/statedb/couchdb/indexes/indexColorSortName.json
[getIndexInfo] indexPath=META-INF/statedb/couchdb/collections/collectionMarbles/indexes/indexCollMarbles.json
[getIndexInfo] indexPath=META-INF/statedb/couchdb/indexColorSortName.json
[getIndexInfo] indexPath=META-INF/statedb/couchdb/collections/indexes/indexCollMarbles.json
[getIndexInfo] indexPath=META-INF/statedb/couchdb/collections/indexCollMarbles.json
[getIndexInfo] indexPath=META-INF/statedb/
--- PASS: TestGetIndexInfo (0.00s)
```

Comment (cendhu) [reply core/ledger/kvledger/txmgmt/privacyenabledstate/db.go]: @mbwhite Existing `TestGetIndexInfo()` has a bug too. It should not pass the path including the index def, i.e., `*.json`. The input to `getIndexInfo()` comes from the `HandleDeployChaincode()` which passes only the directory excluding the actual filename. 

We need to fix

1. `TestGetIndexInfo()`
2. Comments  (provided in the const block)
3. `getIndexInfo()`

Comment (cendhu) [reply core/ledger/kvledger/txmgmt/privacyenabledstate/db.go]: Some tests in `kvledger` package and integration test also fail after making `getIndexInfo()` to consider the full path including the index file.

Comment (mbwhite) [reply core/ledger/kvledger/txmgmt/privacyenabledstate/db.go]: @cendhu ah.. that makes more sense now thanks... what can we say about the directory path that is passed in?
The code as-is is uses '/' to split the path, is it safe to assume the input will not have leading or trailing directory separators?

(aside I assume as well that Fabric is solely intended to work on a Linux filesystem?  / not being the universal file separator :-) )

Comment (cendhu) [reply core/ledger/kvledger/txmgmt/privacyenabledstate/db.go]: @mbwhite 
https://github.com/hyperledger/fabric/blob/7151302be2dff77a9d78cdc8c7a04c4fee352914/core/common/ccprovider/cc_statedb_artifacts_provider.go#L64 This function extracts files entries from the chaincode package (tar bytes). Hence, I assume no leading or trailing directory separator unless there is a bug in `ExtractFileEntries()`. I do not see a unit test for `ExtractFileEntries()`. Not sure whether it is tested somewhere and I miss it. Hence, it is better to verify that. 

As far as I know, even on Windows, we recommend docker to be used (https://hyperledger-fabric.readthedocs.io/en/release-1.4/prereqs.html) and our docker base image is alpine linux. Hence, `/` as a file separator is okay.

Comment (mbwhite) [reply core/ledger/kvledger/txmgmt/privacyenabledstate/db.go]: After a bit more research I believe there are some in db_test.go.  My reading of that function is that it later functions can safely assume there are no leading or trailing /

My background in cross-platform development (pre Docker) has made me question whenever I see anything about a filesystem 'hard coded' like a /

Comment (cendhu) [reply core/ledger/kvledger/txmgmt/privacyenabledstate/db.go]: I see. Thanks for confirming.

**Review by manish-sethi — APPROVED** [review]

**Comment by manish-sethi** [comment]
@Mergifyio backport release-2.2

**Comment by manish-sethi** [comment]
@Mergifyio backport release-2.3

**Pull request merged**